# <center> Lecture10 : Evaluating Regression Models </center>  
 
## <center> Instructor: Dr. Hu Chuan-Peng </center> 

* 在lec9中，我们已经学习了使用一个简单的线性模型来拟合变量间的关系。但这并未结束，接下来，我们需要对模型进行**评估**  

* 我们可以从以下三个角度来思考这个问题：  

    1. 模型本身公正吗？(How fair is the model?)  

    2. 模型存在错误吗？(How wrong is the model?)  

    3. 后验预测模型有多准确？(How accurate are the posterior predictive models?)  


## 模型公正吗(Is the model fair?)  

模型公正性是一个上位概念，它描述了模型是否符合我们(社会、道德、伦理)的预期，而不仅是关注模型和样本数据的关系。  

我们可以借助几个相关问题来理解和思考模型的公众性：  

1. 数据的收集过程是怎样的？  

2. 数据由谁收集，数据收集的目的是什么？  

3. 数据收集的过程，以及分析的结果，将对个人和社会产生什么影响？  

4. 分析过程中可能会包含哪些偏见  
    * p-hacking/HARKing

**结合上节课的例子来思考这几个问题**  

上节课我们发现，压力可以负向预测自我控制。  

* 也就是说，压力越大，自我控制就越差。  

![Image Name](https://cdn.kesci.com/upload/s3y4rmsym1.png?imageView2/0/w/960/h/960)  


**1. 数据的收集的过程公平吗？**  

在本示例研究中，数据来自于被试在问卷上提供的信息。  

* 数据收集过程是公平的，被试自愿完成问卷，并获得相应的报酬。  
* 此外，数据收集过程是匿名化的，保护了被试的隐私。  

一些极端的反例：  
* 在收集数据时，仅考察了大学生被试，而没有考虑在职员工。  
* 在衡量压力大小时，我们通过剥夺被试的睡眠来给被试施加压力。  

🤔基于这些数据的模型公正吗？  

> * 压力量表来源：Cohen, S., Kamarck, T. & Mermelstein, R. A global measure of perceived stress. J. Health. Soc. Behav. 24, 385–396 (1983).  
> * 自我控制量表来源：Tangney, J. P., Baumeister, R. F. & Boone, A. L. High self-control predicts good adjustment, less pathology, better grades, and interpersonal success. J. Pers. 72, 271–324 (2004).

**2. 研究目的，以及数据收集的目的公平吗？**  

在本示例研究中，研究目的来源自心理学家的好奇和假设。这是合理的，因为心理学研究是一种探索性的研究方法。  

一个极端的反例：  
* 如果研究项目来源于开发缓解压力药的厂商。那么，研究目的就可能被操纵，以支持药厂的销售。  
* 例如，有目的的选择本身就具有疾病的患者进行实验。  
* 例如，有目的性的将压力对自我控制的影响夸大，制造消费者的恐慌情绪。

**3. 模型分析的结果，将对个人和社会产生什么影响？**  

在本示例研究中，研究结果(模型分析的结果)具有一定的理论意义和实际意义。  
* 在理论层面，压力对自我控制的影响可能反应了自我功能的特性。  
* 在实际意义层面，这可以为自我控制的干预提供指导。  

一个极端的反例：  
* 例如，在司法领域中工程师开发了一种风险评估和分类模型，该模型可以评估是否拘留还是释放当事人。  
* 即使模型使用的数据是公正的，并且模型在技术上是有效的，但是该模型在几乎在所有情况下都建议拘留当事人。

**4. 分析过程中包含的偏见？**  

在本示例研究中，我们关注一个简单的线性模型，并通过回归系数来反映压力与自我控制的关系。  

一些反例：  
* 假设在一次研究中使用多种问卷收集多种因变量，然后选择有相关性的变量进行报告。  
* 在多因素实验设计中，通过增加变量来获得显著的交互作用，并尝试多种简单效应分析。

此外，在心理学研究中，模型公正性往往与***心理学研究的可重复性***相关。  

1. 数据的收集过程是怎样的？  

2. 数据由谁收集，数据收集的目的是什么？  

3. 数据收集的过程，以及分析的结果，将对个人和社会产生什么影响？  

4. 分析过程中可能会包含哪些偏见  
    * p-hacking/HARKing  

![Image Name](https://pic2.zhimg.com/80/v2-778de6c621356bc2c23c3a09ff2b0be5_1440w.webp)  
(来源：胡传鹏, 王非, 过继成思, 宋梦迪, 隋洁, & 彭凯平. (2016). 心理学研究中的可重复性问题:从危机到契机. 心理科学进展(9), 15.)

## 这个模型可能有多错误(How wrong is the model?)  

> <center>“ all models are wrong, but some are useful.   ————George Box ”</center>  

* 尽管统计模型是对更复杂现实的简化表达，良好的统计模型仍然可以是有用的，并可以增进我们对世界复杂性的理解。  

* 因此，在评估模型时，要问的下一个问题不是模型是否错误(is the model wrong?)，而是模型错误的程度(How wrong is the model?)  

让我们回到lec9的例子中，🤔思考贝叶斯线性回归模型的假设在多大程度上与现实相符？

### 模型假设的影响  

**🤔 我们知道模型存在前提假设，如果这些模型的前提假设不成立，模型会变得多糟糕？**  

在lec9中，我们使用一个线性模型来定义压力分数与自我控制分数之间的关系，并且指定了该模型成立的一些前提假设“  

$$  
\begin{align*}  
Y_i &= \beta_0 + \beta_1 X_i + \epsilon \;\;\;\;\;\;\;\;\epsilon \sim N(0,\sigma^2)\\  
&\Downarrow \\  
Y_i | \beta_0, \beta_1, \sigma &\stackrel{ind}{\sim} N\left(\mu_i, \sigma^2\right) \;\; \text{ with } \;\; \mu_i = \beta_0 + \beta_1X_i  .\\  
\end{align*}  
$$  

* 回归模型需满足如下假设：  

    1. 独立观测假设:每个观测值$Y_i$是相互独立的，即一个观测的值不受其他观测的影响  

    2. 线性关系假设: 预测值$\mu_i$和自变量$X_i$之间可以用线性关系来描述，即：$\mu_i = \beta_0 + \beta_1 X_i$  

    3. 方差同质性假设： 在任意自变量的取值下，观测值$Y_i$都会以$\mu_i$为中心，同样的标准差$\sigma$呈正态分布变化（$\sigma$ is consistent）  

![Image Name](https://cdn.kesci.com/upload/s3y4rmsym1.png?imageView2/0/w/960/h/960)  



**1. 当假设1(独立观测假设)被违反时：**  

* 在心理学的实验数据当中，很多时候观测值之间存在依赖关系(dependent)。  
* 比如，反应时在单个被试内、或某种特定刺激类型内可能更同质：有的被试总是反应得比其他人更快，有的刺激总是能导致更快的反应。  
* 这时，被试内/项目内反应时数据的相互关联会导致对于结果的错误估计。  

* 我们在之后会学习使用层级贝叶斯模型来处理相互关联的数据。  

![Image Name](https://cdn.kesci.com/upload/s3xt2u50rw.png?imageView2/0/w/960/h/960)  

> source: Brown, V. A. (2021). An Introduction to Linear Mixed-Effects Modeling in R. Advances in Methods and Practices in Psychological Science, 4(1), 2515245920960351. https://doi.org/10.1177/2515245920960351

**2. 当假设2(线性关系假设)和假设3(方差同质性假设)被违反时：**  

* 假设2(线性关系假设)违反的情况：在左图中，我们可以看到，Y和X之间的关系并非线性的  

* 假设3(方差同质性假设)违反的情况：并且，随着X的增大，Y的变异性越来越大。  

  * 这要导致的后果是，后验预测分布比实际观测值的分布差异很大(右图)  

![Image Name](https://www.bayesrulesbook.com/bookdown_files/figure-html/nonlinear-1-ch10-1.png)  



### 对模型进行修改  

* 考虑到并不是所有数据都会满足这些假设，当这些假设被违反时，我们需要考虑修改模型。

对于违反假设1 的情况，我们在之后会学习使用层级贝叶斯模型来处理相互关联的数据。  

**对于违反假设2和3的情况，通常有两种处理方式**  

**a. 使用不同的数据模型**  

* 不假设实际值与观测值之间的关系是正态的  $Y_i \sim N(\mu_i, \sigma^2)$  

* 之后，我们会学习使用其他回归模型来描述其数据关系，比如泊松回归、二项回归、负二项回归。  


**b. 对数据进行变换**  

> 如果数据模型并不是我们要担心的问题，我们可以对数据进行变换，仍然可以对变换后的数据可以使用正态模型：  

* 对$Y$进行变换：$g(Y_i) | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2)$，$\mu_i = \beta_0 + \beta_1 X_i$  

* 对$X$进行变换： $Y_i | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2)$，$\mu_i = \beta_0 + \beta_1 h(X_i)$  

* 同时对$Y$和$X$进行变换：$g(Y_i) | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2)$， $\mu_i = \beta_0 + \beta_1 h(X_i)$

在刚刚的例子当中，我们对$Y$的取值做一个对数变换  

$$\log(Y_i) | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2) \;\; \text{ with } \; \mu_i = \beta_0 + \beta_1 X_i$$  

* 在变换之后，可以看到$\log(Y)$与$X$之间的关系仍然是线性的，且随着$X$的增大，$Y$的变异性仍然是一致的。  

* 可以使用正态的线性模型来拟合$\log(Y)$的后验分布  
* 这部分内容将在逻辑回归(logistic regression)部分进行详细介绍，这里先不作深入解释。  

<table>  
        <tr>  
            <td><img src="https://www.bayesrulesbook.com/bookdown_files/figure-html/nonlinear-1-ch10-1.png" alt="" width="500" height="250"></td>  
            <td><img src="https://www.bayesrulesbook.com/bookdown_files/figure-html/nonlinear-2-ch10-1.png" alt="" width="500" height="250"></td>  
        </tr>  
        <tr>  
            <td>变换之前</td>  
            <td>log变换之后</td>  
        </tr>  
</table>  


## 模型评估  

一般情况下，我们的贝叶斯模型不会是完全不公平的，或者错得太离谱的。 但除了这些问题，**更为重要的是，模型是否可以用来准确预测新数据 Y 的结果。**  

> 如果说模型公平性和模型错误描述的是模型在**质量上**的优劣，那模型评估与比较就是在**数量上**衡量模型预测的准确性。  

接下来我们讲介绍衡量后验预测的两种指标：  
- 绝对误差的中位数，median absolute error (MAE)  
- 后验预测区间，posterior prediction interval

### Median absolute error (MAE)  

绝对误差的中位数，median absolute error (MAE)衡量了预测观测值$Y_i$与后验预测均值之间$Y_i'$的差异。  

$$  
\text{MAE} = median\left(|Y_i - \hat{Y}_i|\right)  
$$  

- 假设 $Y_1, Y_2…, Y_n$表示n个观察结果。  

- 每个 $Y_i$ 都有对应的后验预测值，其均值为 $Y_i'$  

结合我们之前的例子来理解绝对误差的中位数：  
- 首先，以压力 $X_i=45$ 下的后验预测值为例。  

- 之后，计算后验预测的均值和观测值之间的误差。  

- 再次，计算所有自变量下，后验预测均值和观测值之间的误差。  

- 最后，计算所有误差的中位数。  

![Image Name](https://cdn-media-1.freecodecamp.org/images/MNskFmGPKuQfMLdmpkT-X7-8w2cJXulP3683)

**（知识回顾）首先，我们考虑压力 $X_i=45$ 下的后验预测模型**  

1. 读取数据 Data_Sum_HPP_Multi_Site_Share，选择"Tsinghua"站点的数据作为示例。  

2. 使用 pymc 对压力变量`stress` 和自我控制变量 `scontrol` 建立回归模型。  

3. 在模型拟合后(略过模型诊断环境)，生成模型的后验预测。  

4. 最后，计算压力 $X_i=45$条件下，后验预测均值和观测值之间的误差。  

*(注：由于压力分数为45时，存在多个自我控制分数与之对应，我们从中随机抽取一个作为演示)*

In [1]:
# 导入 pymc 模型包，和 arviz 等分析工具 
import pymc as pm
import arviz as az
import seaborn as sns
import scipy.stats as st
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr
import pandas as pd

# 忽略不必要的警告
import warnings
warnings.filterwarnings("ignore")

In [3]:
# 通过 pd.read_csv 加载数据 Data_Sum_HPP_Multi_Site_Share.csv
df_raw = pd.read_csv('/home/mw/input/bayes20238001/Data_Sum_HPP_Multi_Site_Share.csv')

# 筛选站点为"Tsinghua"的数据
df = df_raw[df_raw["Site"] == "Tsinghua"]

# 选择需要的两列
df = df[["stress","scontrol"]]

#设置索引
df["index"] = range(len(df))
df = df.set_index("index")

In [4]:
# 从压力值为45对应的自我控制分数中随机抽取一个，并标记其索引号
# 我们将根据该索引号来提取对应的后验预测值
print(df[df.stress == 45].sample(n=1, random_state=84735))

       stress  scontrol
index                  
123        45        35


In [7]:
# 对压力变量`stress` 和自我控制变量 `scontrol` 建立回归模型
with pm.Model() as linear_model:

    beta_0 = pm.Normal("beta_0", mu=70, sigma=4)          #定义beta_0          
    beta_1 = pm.Normal("beta_1", mu=0, sigma=0.8)         #定义beta_1
    sigma = pm.Exponential("sigma", 0.6)                  #定义sigma

    #x是自变量压力水平
    x = pm.MutableData("x",df.stress,dims="obs_id")       
    #定义mu，讲自变量与先验结合
    mu = pm.Deterministic("mu", beta_0 + beta_1 * x, dims="obs_id")    
    #定义似然：预测值y符合N(mu, sigma)分布, observed 为传入实际数据y 自我控制水平
    likelihood = pm.Normal("y_est", mu=mu, sigma=sigma, observed=df.scontrol, dims="obs_id")   

In [8]:
# 模型拟合 
#===========================
#     注意！！！以下代码可能需要运行1-2分钟左右
#===========================
with linear_model:
    trace = pm.sample(draws=5000,                   # 使用mcmc方法进行采样，draws为采样次数
                      tune=1000,                    # tune为调整采样策略的次数，可以决定这些结果是否要被保留
                      chains=4,                     # 链数
                      discard_tuned_samples= True,  # tune的结果将在采样结束后被丢弃
                      random_seed=84735)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_1, sigma]


Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 58 seconds.


In [9]:
# 进行后验预测采样
with linear_model:
    ppc = pm.sample_posterior_predictive(trace,
                                         random_seed=84735)


Sampling: [y_est]


**计算预测误差(prediction error)**  

接下来，我们从`ppc`中提取压力 $X_i=45$条件下的生成的后验预测，并计算后验预测的均值和观测值之间的误差(prediction error)。  

- 查看 `ppc`，可以发现其中存储了`posterior_predictive`，`observed_data`，`constant_data`  
    - 生成的后验预测存储在`posterior_predictive`中  

    - `observed_data`对应原始数据集中的因变量，自我控制分数(df.scontrol)  
    
    - `constant_data`对应原始数据集中的自变量，压力分数(df.scontrol)  

- 之前提取出的自变量$X_i=45$，其对应的索引为123，我们将利用这个索引提取存储在`ppc`中对应的真实值、对应的后验预测值

In [10]:
ppc

Inference data with groups:
	> posterior_predictive
	> observed_data
	> constant_data

In [11]:
# 确定constant_data(即自变量)中索引值为123的压力分数确实是 45
ppc.constant_data["x"][123].values == 45

True

In [12]:
# 提取压力分数 45对应的后验预测值
ppc_45 = ppc.posterior_predictive["y_est"][:,:,123]

In [13]:
ppc_45

<xarray.DataArray 'y_est' (chain: 4, draw: 5000)>
array([[24.9845489 , 35.98246459, 34.77980758, ..., 41.90577622,
        34.11042392, 42.99840913],
       [32.14299208, 38.39322657, 35.07711454, ..., 44.11957206,
        25.18009865, 28.53782086],
       [38.11492019, 36.49526697, 43.27348308, ..., 45.54957557,
        41.68482094, 36.92279602],
       [39.52245511, 37.99750227, 37.47455231, ..., 39.20713261,
        32.44799684, 29.73620739]])
Coordinates:
  * chain    (chain) int64 0 1 2 3
  * draw     (draw) int64 0 1 2 3 4 5 6 7 ... 4993 4994 4995 4996 4997 4998 4999
    obs_id   int64 123

**计算后验预测的均值:**  

- 提取所有 链(chain:4)和 采样次数(4*5000)下，的后验预测值  

- 计算平均值  

- 结果显示，在压力 $X_i=45$时，后验预测(自我控制得分)的均值为36.9

In [14]:
ppc_45_mean = ppc_45.stack(sample=("chain", "draw")).mean().values

print("prediction error:", ppc_45_mean)

prediction error: 36.92569395480895


**计算后验预测均值和观测值之间的误差**  

我们看一下在压力 $X_i=45$时，原数据中的自我控制分数  

然后，进一步计算后验预测误差(prediction error)。  
- $\text{prediction error} = Y-Y'$  

- 即观测值$(Y)$和后验预测均值$(Y')$之间的误差。

In [15]:
print("Observed Value(X=45):",ppc.observed_data["y_est"][123].values)

Observed Value(X=45): 35.0


In [16]:
# 提取观测数据
true_y = ppc.observed_data["y_est"][123].values

# 获得预测模型的均值
pre_y_mean = ppc_45_mean

# 计算后验预测误差
pre_error = true_y - pre_y_mean
print("prediction error:", pre_error)

prediction error: -1.9256939548089491


**计算所有观测变量下后验预测模型的均值和观测值之间的误差**  



$$  
\begin{array}{lll}  
\text{Markov chain parameter sets} & & \text{Simulated posterior predictive models} \\  
\left[  
\begin{array}{lll}  
\beta_0^{(1)} & \beta_1^{(1)} & \sigma^{(1)} \\  
\hline  
\beta_0^{(2)} & \beta_1^{(2)} & \sigma^{(2)} \\  
\hline  
\vdots & \vdots & \vdots \\  
\hline  
\beta_0^{(20000)} & \beta_1^{(20000)} & \sigma^{(20000)} \\  
\end{array}  
\right]  
& \; \longrightarrow \;  
& \left[  
\begin{array}{l|l|l|l}  
Y_1^{(1)} & Y_2^{(1)} & \cdots & Y_{174}^{(1)} \\  
Y_1^{(2)} & Y_2^{(2)} & \cdots & Y_{174}^{(2)} \\  
\vdots & \vdots &  & \vdots \\  
Y_1^{(20000)} & Y_2^{(20000)} & \cdots & Y_{174}^{(20000)} \\  
\end{array}  
\right]  
\end{array}  

$$  

* 在当前数据集中，自变量共有`174`个  

* 因此，相对于之前我们仅考虑 $X_i=45$ 时的单一情况，现在我们可以计算 所有$X$取值下对应的后验预测的均值。  

  * 代入`20000`个参数对，每个自变量都拥有对应的20000个后验预测值  

  * 对于每个$X$，计算后验预测的均值  

  * 进而，计算每个$X$下，后验预测均值和观测值之间的误差  

![Image Name](https://cdn-media-1.freecodecamp.org/images/MNskFmGPKuQfMLdmpkT-X7-8w2cJXulP3683)  

资料来源：https://www.freecodecamp.org/news/machine-learning-mean-squared-error-regression-line-c7dde9a26b93/

In [17]:
ppc.posterior_predictive["y_est"]

<xarray.DataArray 'y_est' (chain: 4, draw: 5000, obs_id: 174)>
array([[[37.17928622, 26.98226507, 49.90507747, ..., 29.15132464,
         39.83179059, 41.79911201],
        [49.17884201, 40.60248473, 42.92529399, ..., 26.46296462,
         37.95914918, 43.04134461],
        [44.12856608, 36.273458  , 55.21917594, ..., 38.07751015,
         43.72080748, 41.5922247 ],
        ...,
        [30.81905244, 35.56341664, 45.78583365, ..., 33.31537037,
         40.94832677, 40.8900699 ],
        [44.85905922, 35.96239992, 48.73922393, ..., 34.93553305,
         39.08652819, 45.05557962],
        [46.99660879, 42.43470429, 47.46189327, ..., 45.89091776,
         27.73549683, 37.0572504 ]],

       [[49.36445521, 37.93724032, 41.49223021, ..., 28.78552014,
         29.54991112, 46.77650779],
        [46.66767795, 46.51453661, 46.38961224, ..., 38.85276683,
         46.44636293, 44.84504182],
        [38.62356186, 50.69913222, 50.23194441, ..., 44.60292732,
         43.31584756, 45.0553892 ],
...
        [42.72839636, 43.28034965, 43.0241022 , ..., 34.5373706 ,
         43.2713028 , 43.79617922],
        [43.66511777, 40.7236543 , 49.89061319, ..., 43.06994767,
         42.63813397, 41.26049574],
        [37.09684265, 39.64349381, 41.00761416, ..., 32.67667397,
         46.71104257, 39.75396418]],

       [[42.78286292, 38.34838403, 44.22095517, ..., 36.84721693,
         39.03293796, 54.33830171],
        [48.82541634, 26.88475394, 47.36051818, ..., 40.64579572,
         37.63592973, 44.12419868],
        [43.81625484, 38.86094454, 41.41020442, ..., 33.96155172,
         45.98649695, 34.58731663],
        ...,
        [33.53733419, 34.36293246, 42.45918396, ..., 32.62125852,
         43.69755566, 45.52457825],
        [33.58336838, 45.13866273, 51.22329994, ..., 38.77243183,
         37.31382425, 52.67308152],
        [32.72375452, 41.73460936, 40.29641465, ..., 33.43859207,
         37.83074102, 43.26893527]]])
Coordinates:
  * chain    (chain) int64 0 1 2 3
  * draw     (draw) int64 0 1 2 3 4 5 6 7 ... 4993 4994 4995 4996 4997 4998 4999
  * obs_id   (obs_id) int64 0 1 2 3 4 5 6 7 ... 166 167 168 169 170 171 172 173

In [18]:
# 提取出所有的后验预测值，每一个X对应4*5000个后验预测值，使用stack汇总到一起
pre_per_x = ppc.posterior_predictive["y_est"].stack(sample=("chain", "draw"))

#计算均值
pre_y_mean = pre_per_x.mean(axis=1).values

# 提取自变量X，对应的观测值Y、后验预测均值，存放在一个表格中
pre_y_mean = pd.DataFrame({
    "stress": df.stress,
    "scontrol_ppc_mean": pre_y_mean,
    "scontrol_original": df.scontrol
})

#显示表格前10行
pre_y_mean.head(10)

stress  scontrol_ppc_mean  scontrol_original
index                                              
0          38          40.462623                 47
1          43          37.982065                 40
2          31          44.036412                 40
3          35          42.084572                 46
4          45          36.946839                 50
5          42          38.466323                 43
6          41          38.958656                 36
7          58          30.366001                 35
8          47          35.925984                 47
9          33          43.089108                 50

**计算所有误差的中位数(MAE)**  

$$  
{\mathsf{MAE}}={\mathsf{median}}|Y_{i}-Y_{i}^{\prime}|.  
$$  

- MAE 为所有预测误差Y-Y'的中位数。  

最后，我们计算所有观测值与后验预测模型均值之间的误差。

In [19]:
from statistics import median

# 计算预测误差
pre_y_mean["pre_error"] = pre_y_mean["scontrol_original"] - pre_y_mean["scontrol_ppc_mean"]

# 最后，计算预测误差的中位数
MAE = median(pre_y_mean.pre_error)
print(f"计算得到的MAE: {MAE:.2f}")

计算得到的MAE: -0.56


In [20]:
# 可视化预测误差MAE
pre_error = pre_y_mean.pre_error.values
az.plot_posterior(pre_error, hdi_prob=0.95, point_estimate="median", ref_val=MAE)

<Axes: title={'center': 'x'}>

<Figure size 640x480 with 1 Axes>

对于 MAE 的解读。  

- 绝对误差的中位数，median absolute error (MAE)衡量了预测观测值$Y_i$与后验预测均值之间$Y_i'$的差异。  
- **MAE越小**表明后验模型的**预测越准确**。  



### 后验预测区间，posterior prediction interval  

- MAE 是一个有效的指标，可以衡量模型对于样本观测值的预测能力。  

- 然而，MAE 仅考虑了后验预测误差的中位数，而不是其可信区间。  

- 后验预测区间(posterior prediction interval)则进一步反应了样本观测值与后验预测可信区间的关系。  

![Image Name](https://cdn.kesci.com/upload/s3xt34m35o.png?imageView2/0/w/700/h/700)

为了便于理解，我们还是回到仅考虑$X_i=45$的情况。  

- MCMC采样针对该$X_i$生成了20000个后验预测值  
- 我们可以用`az.summary`快速查看这20000个后验预测值的情况，包括均值和可信区间  

此时, 后验预测的均值为36  
- 并且，此时的94%可信区间为27到47.7  
- 对比真实数据可以发现，有2个观测值(26、50)落在可信区间外。

In [21]:
az.summary(ppc_45)

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  ess_tail  \
y_est  36.926  5.508  26.947   47.707      0.039    0.028   19497.0   18710.0   

       r_hat  
y_est    1.0

In [22]:
df.query("stress == 45")

stress  scontrol
index                  
4          45        50
12         45        26
103        45        31
123        45        35
127        45        36
149        45        34

**计算所有X下，观测值落在后验预测可信区间的情况**  



$$  
\begin{array}{lll}  
\text{Markov chain parameter sets} & & \text{Simulated posterior predictive models} \\  
\left[  
\begin{array}{lll}  
\beta_0^{(1)} & \beta_1^{(1)} & \sigma^{(1)} \\  
\hline  
\beta_0^{(2)} & \beta_1^{(2)} & \sigma^{(2)} \\  
\hline  
\vdots & \vdots & \vdots \\  
\hline  
\beta_0^{(20000)} & \beta_1^{(20000)} & \sigma^{(20000)} \\  
\end{array}  
\right]  
& \; \longrightarrow \;  
& \left[  
\begin{array}{l|l|l|l}  
Y_1^{(1)} & Y_2^{(1)} & \cdots & Y_{174}^{(1)} \\  
Y_1^{(2)} & Y_2^{(2)} & \cdots & Y_{174}^{(2)} \\  
\vdots & \vdots &  & \vdots \\  
Y_1^{(20000)} & Y_2^{(20000)} & \cdots & Y_{174}^{(20000)} \\  
\end{array}  
\right]  
\end{array}  

$$  

* 在当前数据集中，我们可以计算每一个X取值下，对应的后验预测可信区间。  

  * 代入`20000`个参数对，每个自变量都拥有对应的`20000`个后验预测值，通过这些值获得可信区间  

* 计算出`174`个可信区间之后，并观察真实值落在可信区间的情况  


**计算后验预测HDI**  

我们先计算单个X对应的后验预测HDI，并打包为函数：  

- 首先，定义函数，计算对应X下的后验预测HDI范围  

- 其次，判断观测值是否超过 HDI 范围  

- 最后，记录并返回超过 HDI 范围的观测值的数量

In [23]:
ppc

Inference data with groups:
	> posterior_predictive
	> observed_data
	> constant_data

In [24]:
# 定义函数，计算对应X下的观测值超出后验预测模型hdi范围的数量
def count_outlier(i,hdi=0.94,ppc=ppc):
  # 提取第i个X的值
  X = ppc.constant_data["x"][i].values
  # 提取第i个X值对应的后验预测值
  pre_ys = ppc.posterior_predictive["y_est"].stack(sample=("chain", "draw"))[i].values
  # 提取对应X值下的观测值Y
  true_ys = ppc.observed_data["y_est"][i].values
  
  # 计算对应后验预测的hdi的上下限
  lower,upper = az.hdi(pre_ys, hdi_prob=hdi)
  
  # 判断是否超过 hdi 边界
  outlier = (true_ys > upper) | (true_ys < lower)

  return outlier

# 判断X为45时，对应的观测值Y是否超出后验预测HDI范围
i = 123
print(f"X为{df.stress[i]}时，Y是否超出后验预测HDI:", count_outlier(i))

X为45时，Y是否超出后验预测HDI: False


* 通过一次定义，我们可以得到一个$X$下，其对应的观测值$Y$是否超出后验预测HDI。  

* 我们可以通过循环，对每个$X$执行这个操作，再总结$Y$超出后验预测HDI的次数。

In [31]:
# 建立一个list,存储是否超过后验预测范围的判断结果
hdi_verify = []

# 使用循环，判断所有X对应的Y是否超出后验预测范围，并计数
for i in range(len(df)):  
  hdi_verify.append(count_outlier(i, hdi=0.94))
total_number = sum(hdi_verify)


print(f"所有超过后验预测范围HDI的数量: {total_number}")

所有超过后验预测范围HDI的数量: 11


* 考虑到我们一共有 174 个观测值，其中有11个观测值超过94%的后验预测HDI  

* 这表明，我们的预测准确率超过 90%，$(174-11)/174 = 0.937$

In [26]:
# 可视化后验预测区间的结果
az.plot_lm(idata=ppc,           #传入后验预测的数据集
           x='x',               #传入ppc中，自变量的名称
           y='y_est',           #传入ppc中，观测变量的名称
           kind_pp="hdi",       #绘制可信区间
           kind_model="hdi")
plt.legend(loc="upper right")

<Figure size 640x480 with 1 Axes>

In [32]:
# 建立一个list,存储是否超过后验预测范围的判断结果
hdi_verify = []

##================================================
#                练习，观察不同HDI范围(例如hdi=50%)下，超出后验预测范围观测值的数量
#                提示：通过向count_outlier中加入 hdi=0.5可以调整HDI的范围为50%
#================================================

# 使用循环，判断所有X对应的Y是否超出后验预测范围，并计数
for i in range(...):  
  hdi_verify.append(...)
total_number = sum(...)


print(f"所有超过后验预测范围观测值的数量: {total_number}")

In [ ]:
# 可视化后验预测区间的结果
az.plot_lm(idata=ppc,           #传入后验预测的数据集
           x='x',               #传入ppc中，自变量的名称
           y='y_est',           #传入ppc中，观测变量的名称
           kind_pp="hdi",       #绘制可信区间
           y_hat_fill_kwargs={"hdi_prob":0.5},
           kind_model="hdi")
plt.legend(loc="upper right")

## 练习: 当自变量为二分离散变量  

* 我们使用一个新例子来巩固对**MAE**与**后验预测区间**的计算  

* 在上一个例子中，我们以连续变量 *压力* 作为自变量预测*自我控制分数*  

* 类别变量同样也可以作为自变量  

* 这里，我们考虑一个简单的二分变量*是否吸烟*，将其作为自变量，来预测*自我控制分数*  
	* 我们可以有个直觉的猜测：比如吸烟的个体自控力可能更差。

In [33]:
# 加载数据 Data_Sum_HPP_Multi_Site_Share.csv
df_raw = pd.read_csv('/home/mw/input/bayes20238001/Data_Sum_HPP_Multi_Site_Share.csv')

# 筛选站点为"Tsinghua"的数据
df_METU = df_raw[df_raw["Site"] == "METU"]

# 选择需要的两列
df_METU = df_METU[["stress","scontrol","smoke"]]

#1 表示抽烟，2表示不抽烟---重新编码---0 表示不抽烟，1表示抽烟
df_METU["smoke"] =  np.where(df_METU['smoke'] == 2, 0, 1)
df_METU["smoke_recode"] =  np.where(df_METU['smoke'] == 1, "yes", "no")

#设置索引
df_METU["index"] = range(len(df_METU))
df_METU = df_METU.set_index("index")

* 在这个例子中，我们将是否吸烟作为自变量，将自我控制分数作为因变量  

    * 在自变量中，`0` 表示不吸烟，`1`表示吸烟  

$$  
X_{i2} = \begin{cases}  
1 & \text{Yes} \\  
0 & \text{No} \\  
\end{cases}  
$$  

简单画出两种情况下，自我控制分数分布的密度图  
* 可见不吸烟的个体的自我控制得到分高于吸烟个体。  

![Image Name](https://cdn.kesci.com/upload/s40nk3kagu.png?imageView2/0/w/960/h/960)

* 同样地，我们可以为抽烟和自我控制之间的关系建立一个正态回归模型：  

$$\begin{equation}  
\begin{array}{lcrl}  
\text{data:} & \hspace{.05in} & Y_i|\beta_0,\beta_1,\sigma & \stackrel{ind}{\sim} N(\mu_i, \; \sigma^2) \;\; \text{ with } \;\; \mu_i = \beta_0 + \beta_1 X_{i2}\\  
\text{priors:} & & \beta_{0}  &  \sim N\left(60, 2^2\right)  \\  
                     & & \beta_1  & \sim N\left(-10, 4^2 \right) \\  
                     & & \sigma & \sim \text{Exp}(0.6)  .\\  
\end{array}  
\end{equation}  
$$  

**在这个例子中，各参数的意义：**  

* $\beta_0  =  \beta_0 + \beta_1 \cdot 0$  

    * 当$X_{i2} = 0$时，$mu_i$的取值。 $\beta_0$ 表示不抽烟情况下，自我控制分数的均值  

* $\beta_0 + \beta_1 \cdot 1 = \beta_0 + \beta_1$  

    * $\beta_1 $ ：当$X_{i2} = 1$时，$mu_i$的取值。表示抽烟情况下，自我控制分数的均值，则$\beta_1 $表示了从不抽烟到抽烟，自我控制分数的变化程度  

* 这里我们还是根据经验定义先验中的超参

**接下来我们直接运行模型定义、MCMC采样、后验预测部分**

In [34]:
with pm.Model() as category_model:

    beta_0 = pm.Normal("beta_0", mu=60, sigma=2)          #定义beta_0          
    beta_1 = pm.Normal("beta_1", mu=-10, sigma=4)         #定义beta_1
    sigma = pm.Exponential("sigma", 0.6)                  #定义sigma

    smoke = pm.MutableData("smoke",df_METU.smoke, dims="obs_id")
    mu = pm.Deterministic("mu", beta_0 + beta_1*smoke, dims="obs_id")                           #定义mu，讲自变量与先验结合

    likelihood = pm.Normal("y_est", mu=mu, sigma=sigma, observed=df_METU.scontrol, dims="obs_id")   #定义似然：预测值y符合N(mu, sigma)分布
                                                                                #通过 observed 传入实际数据y 自我控制水平

In [35]:
#===========================
#     注意！！！以下代码可能需要运行1-2分钟左右
#===========================
with category_model:
    cat_regression = pm.sample(draws=5000,                   # 使用mcmc方法进行采样，draws为采样次数
                      tune=1000,                    # tune为调整采样策略的次数，可以决定这些结果是否要被保留
                      chains=4,                     # 链数
                      discard_tuned_samples= True,  # tune的结果将在采样结束后被丢弃
                      random_seed=84735)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_1, sigma]


Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 26 seconds.


In [36]:
with category_model:
    cat_ppc = pm.sample_posterior_predictive(cat_regression,
                                             random_seed=84735) 

Sampling: [y_est]


### 代码练习  

* 我们以这个新模型为例，进行对MAE、后验预测区间代码的完整练习  

* 结合之前的代码，对缺失的部分进行补充

In [37]:
cat_ppc

Inference data with groups:
	> posterior_predictive
	> observed_data
	> constant_data

### 计算MAE：  

* 温馨提示：仔细查看上一行的`cat_ppc`中的各个变量，确保自己大致理解其中的含义

In [ ]:
##================================================
#                练习，修改... 部分
#                
#================================================

# 计算每个X取值下对应的后验预测模型的均值
cat_pre_x = cat_ppc.posterior_predictive[...].stack(sample=("chain", "draw"))
cat_pre_y_mean = cat_pre_x.mean(axis=1).values

# 提取对应X值下的观测值Y，提取对应X值下的后验预测模型的均值
cat_MAE = pd.DataFrame({
    "smoke": ...,
    "scontrol_ppc_mean": ...,
    "scontrol_original": ...
})
cat_MAE.head(10)

In [ ]:
from statistics import median

# 计算预测误差
cat_MAE["pre_error"] = cat_MAE[...] - cat_MAE[...]

# 最后，计算预测误差的中位数
MAE = median(cat_MAE.pre_error)
print(f"计算得到的MAE: {MAE:.2f}")

### 计算后验预测区间

In [38]:
cat_ppc

Inference data with groups:
	> posterior_predictive
	> observed_data
	> constant_data

In [ ]:
# 定义函数，计算对应X下的观测值超出后验预测模型hdi范围的数量
def count_outlier(i,hdi=0.94,ppc=cat_ppc):
  # 提取第i个X的值
  X = ppc.constant_data["smoke"][i].values
  # 提取第i个X值对应的后验预测值
  pre_ys = ppc.posterior_predictive["y_est"].stack(sample=("chain", "draw"))[i].values
  # 提取对应X值下的观测值Y
  true_ys = ppc.observed_data["y_est"][i].values
  
  # 计算对应后验预测的hdi的上下限
  lower,upper = az.hdi(pre_ys, hdi_prob=hdi)
  
  # 判断是否超过 hdi 边界
  outlier = (true_ys > upper) | (true_ys < lower)

  return outlier

In [ ]:
# 建立一个list,存储是否超过后验预测范围的判断结果
hdi_verify = []

# 使用循环，判断所有X对应的Y是否超出后验预测范围，并计数
for i in range(len(df_METU)):  
  hdi_verify.append(count_outlier(...))
total_number = sum(...)

print(f"所有超过后验预测范围观测值的数量: {total_number}")

In [39]:
cat_ppc

Inference data with groups:
	> posterior_predictive
	> observed_data
	> constant_data

In [ ]:
az.plot_lm(idata=...,           #传入后验预测的数据集
           x="...",             #传入ppc中，自变量的名称
           y="...",             #传入ppc中，观测变量的名称
           kind_pp="hdi",       #绘制可信区间
           kind_model="hdi",
           y_kwargs={"alpha":0.5})
plt.legend(loc='upper right')
plt.xticks([0,1],["no","yes"],fontsize=15)
plt.yticks(fontsize=15)
plt.ylabel("Self control",fontsize=16)
plt.xlabel("Smoke",fontsize=16)

## 总结  



本节课讨论了模型的公平性和模型错误程度的话题。更重要的是结合后验预测检查和后验预测准确性来判断模型的预测效果。  
       
- 为了确定模型的后验预测准确性，我们计算两个重要的指标：  
    
    - 绝对误差的中位数，median absolute error (MAE)  
    - 后验预测区间(posterior prediction interval)  

- 此外，我们介绍了自变量为离散变量的回归模型，并进行了练习。  
	- 🤔一个值得思考的问题是：对于吸烟的个体，他们的压力是否更大，是否因为压力更大而导致自我控制力低？  
	- 下节课我们将探索这个问题。